sdk/nodejs/runtime/closure/v8_v11andHigher.ts

```diff
@@ -0,0 +1,214 @@
+    }
+
+    if (!scopes.value.objectId) {
+        throw new Error("[[Scopes]].value have objectId");
+    }
+
+    // This is sneaky, but we can actually map back from the [[Scopes]] object to a real in-memory
+    // v8 array-like value.  Note: this isn't actually a real array.  For example, it cannot be
+    // iterated.  Nor can any actual methods be called on it. However, we can directly index into
+    // it, and we can.  Similarly, the 'object' type it optionally points at is not a true JS
+    // object.  So we can't call things like .hasOwnProperty on it.  However, the values pointed to
+    // by 'object' are the real in-memory JS objects we are looking for.  So we can find and return
+    // those successfully to our caller.
+    const scopesArray: { object?: Record<string, any> }[] = await getValueForObjectId(scopes.value.objectId);
+
+    // scopesArray is ordered from innermost to outermost.
+    for (let i = 0, n = scopesArray.length; i < n; i++) {
+        const scope = scopesArray[i];
+        if (scope.object) {
+            if (freeVariable in scope.object) {
+                const val = scope.object[freeVariable];
+                return val;
+            }
+        }
+    }
+
+    if (throwOnFailure) {
+        throw new Error("Unexpected missing variable in closure environment: " + freeVariable);
+    }
+
+    return undefined;
+}
+
+// In order to get information about an object, we need to put it in a well known location so that
+// we can call Runtime.evaluate and find it.  To do this, we just make a special map on the 'global'
+// object, and map from a unique-id to that object.  We then call Runtime.evaluate with an
+// expression that then points to that unique-id in that global object.  The runtime will then find
+// the object and give us back an internal id for it.  We can then query for information about the
+// object through that internal id.
+//
+// Note: the reason for the mapping object and the unique-id we create is so that we don't run into
+// any issues when being called asynchronously.  We don't want to place the object in a location
+// that might be overwritten by another call while we're asynchronously waiting for our original
+// call to complete.
+
+let currentFunctionId = 0;
+// tslint:disable-next-line:variable-name
+const __inflightFunctions: any = {};
+(<any>global).__inflightFunctions = __inflightFunctions;
+
+async function getRuntimeIdForFunctionAsync(func: Function): Promise<inspector.Runtime.RemoteObjectId> {
+    // Place the function in a unique location off of the global object.
+    const currentFunctionName = "id" + currentFunctionId++;
+    __inflightFunctions[currentFunctionName] = func;
+
+    try {
+        const session = await v8Hooks.getSessionAsync();
+
+        const expression = `global.__inflightFunctions.${currentFunctionName}`;
+        const retType = await new Promise<inspector.Runtime.EvaluateReturnType>((resolve, reject) => {
```

Comment:
You can use `util.promisify` if you don't like having to create new promises like this - it's just some sugar on top of this pattern.

Reply:
https://nodejs.org/dist/latest-v8.x/docs/api/util.html#util_util_promisify_original

Reply:
made this work.  albeit painfullly.  There are several TS issues that make this not as nice as we would like :)